demos/bounded_heat/bounded_heat.py.rst

```diff
@@ -0,0 +1,202 @@
+Solving the Heat Equation with Bounds Constraints
+=================================================
+
+In this demo we solve the simple heat equation with bounds constraints applied uniformly in time and space. 
+Consider the simple heat equation on :math:`\Omega = [0,1]\times [0,1]`, with boundary :math:`\Gamma`:
+
+.. math::
+
+    u_t - \Delta u &= f
+
+    u & = g \quad \textrm{on}\ \Gamma
+
+for some known functions :math:`f` and :math:`g`. The solution will be some function :math:`u\in V`, for 
+a suitable function space :math:`V`.
+
+The weak form is found by multiplying by an arbitrary test function :math:`v\in V` and integrating over :math:`\Omega`. 
+We then have the variational problem of finding :math:`u:[0,T]\rightarrow V` such that 
+.. math::
+
+    (u_t, v) + (\nabla u, \nabla v) = (f, v)
+
+This demo uses particular choices of the functions :math:`f` and :math:`g` to be defined below.
+
+First, we must import firedrake and certain items from Irksome: ::
+
+    from firedrake import *
+    from irksome import (Dt, MeshConstant, RadauIIA, TimeStepper)
+
+We also need some UFL tools in order to manufacture a solution: ::
+
+    from ufl.algorithms import expand_derivatives
+
+Finally, numpy provides us with the upper bound of infinity: ::
+
+    import numpy as np
+
+We first define the mesh and the necessary function space. We choose 
+quadratic Bernstein polynomials to support bounds-constraints in space: ::
+
+    N = 32
+
+    msh = UnitSquareMesh(N, N)
+    V = FunctionSpace(msh, "Bernstein", 2)
+
+In order to enforce bounds constraints in time, we must utilize a collocation method. 
+In this demo, we will time-step using the L-stable, fully implicit, 2-stage RadauIIA 
+Runge-Kutta method. The bounds will be passed as an argument to the 
+:meth:`~.TimeStepper.advance` method. We now define the Butcher Tableau and variables to store the 
+time step, current time, and final time: ::
+
+    butcher_tableau = RadauIIA(2)
+
+    MC = MeshConstant(msh)
+    dt = MC.Constant(2 / N)
+    t = MC.Constant(0.0)
+    Tf = MC.Constant(1.0)
+
+We will find an approximate solution at time :math:`t=1.0` with and without 
+enforcing a constraint on the lower bound. We will need the following pair of solver 
+parameters: ::
+
+    lu_params = {
+        "snes_type": "ksponly",
+        "ksp_type": "preonly",
+        "mat_typ": "aij",
```

Comment:
There is a typo here: `mat_type`.  Also, for parallel structure with `vi_params`, we should set it in both places or neither.